cdap-docs/developers-manual/source/advanced/class-loading.rst

```diff
@@ -0,0 +1,100 @@
+- Different plugins can choose to use any library of any version, yet all plugins are
+  usable by the same application at the same time.
+
+Parent Delegation Model
+=======================
+
+When a Java Virtual Machine (JVM) needs to load a class, by default it follows the
+`parent delegation model 
+<https://www-01.ibm.com/support/knowledgecenter/#!/SSYKE2_7.0.0/com.ibm.java.zos.70.doc/diag/understanding/cl_delegation.html>`__. 
+The parent delegation model allows for one class to be loadable from multiple class
+loaders, as long as those class loaders have the defining class loader of that class as a
+common ancestor. This property enables one to define an extensible class loading
+architecture that provides class isolation, yet at the same time they’re able to
+inter-operate with a set of common API classes.
+
+Class Loading in CDAP
+=====================
+Based on the parent delegation model, we’ve come up with this class loader hierarchy in
+CDAP:
+
+.. image:: ../_images/class-loading/class-loading01.png
+   :align: center
+
+..    :width: 82%
+..    :scale: 75%
+
+With this class loader hierarchy, the CDAP runtime system interacts with custom
+applications only through classes and interfaces defined in the API class loader; hence,
+libraries used by the CDAP system itself won’t interfere with applications. 
+
+Moreover, CDAP allows an application to define its own plugin API for extending the
+functionalities of the application itself. Each plugin implementation will be loaded
+through a separate plugin class loader. The parent of each plugin class loader is a filter
+class loader that ensures that only CDAP and plugin API classes are loaded from the
+application class loader. All other classes required by the plugin are loaded by the
+plugin class loader itself. This allows an application to use multiple plugins at the same
+time, while providing class loading isolation between different plugins.
+
+In CDAP, both applications and plugins are deployed as JAR files to the system. Instead of
+a normal JAR file that contains a collection of class files, a bundle JAR file has content
+similar to this example:
+
+.. image:: ../_images/class-loading/class-loading02.png
+   :align: center
+
+..    :width: 58%
+..    :scale: 75%
+
+Inside an application bundle JAR file, besides the class files for the application, are
+the JAR files for third party libraries that are needed by the application. In the
+application JAR manifest, a special attribute (“Export-Package”) can be added to instruct
+the filter class loader which packages are contained in the application plugin API. Hence,
+only classes under those packages will be available to the plugin. This is to avoid
+libraries used by the application interfering with the ones used by the plugin. The plugin
+JAR file is following a similar structure as an application JAR file.
+
+Class Loading in Hadoop MapReduce and Apache Spark
```

Comment:
seems a little inconsistent to use Apache for Spark but not for Hadoop. Maybe use Apache Hadoop MapReduce and Apache Spark in the title, and only MapReduce and Spark in the content below.